packages/svelte-urql/src/common.ts

```diff
@@ -0,0 +1,136 @@
+ *
+ * This probably shouldn't be a Svelte-specific interface, however `@urql/core` doesn't appear to define it
+ *
+ * @todo
+ * preferably, this would also have status properties for convenience:
+ *
+ * - `isIdle`: !fetching && !data
+ * - `isFetching`: !!fetching && !data
+ * - `isSuccess`: !error && !!data
+ * - `isError`: !!error
+ */
+export interface AnnotatedOperationResult<Data, Variables>
+  extends OperationResult<Data, Variables> {
+  /** `true`: awaiting fetch response (no error, no data), otherwise `false` */
+  fetching: boolean;
+}
+
+/** Options passed to `queryStore`, `mutationStore`, and `subscriptionStore` */
+export interface UrqlStoreArgs<Data, Variables extends object = {}> {
+  /** an [Urql client](https://formidable.com/open-source/urql/docs/api/core/#client)  */
+  client: Client;
+  /** a graphql tag */
+  query: DocumentNode | TypedDocumentNode<Data, Variables> | string;
+  /** the variables to be used in the fetch operation */
+  variables: Variables;
+  /** Urql fetching options */
+  context?: Partial<OperationContext>;
+  /** Convenience input.  Ignored if context.requestPolicy is provided */
+  requestPolicy?: RequestPolicy;
+}
+
+/**
+ * A Svelte ["Readable" store](https://svelte.dev/docs#run-time-svelte-store-readable) containing Urql [operation results](https://formidable.com/open-source/urql/docs/api/core/#operationresult) along with some convenience tools.
+ */
+export type UrqlStore<Data, Variables extends object = {}> = Readable<
+  AnnotatedOperationResult<Data, Variables>
+>;
+
+/**
+ * Make a Wonka source from a Svelte store
+ * @param store$ a Svelte store
+ * @returns a [Wonka source](https://wonka.kitten.sh/api/sources#makesubject) subscribing to the store's changes
+ */
+export function fromStore<T>(store$: Readable<T>) {
+  const subject = makeSubject<T>();
+  const { source, next } = subject;
+  store$.subscribe(value => next(value));
+  return source;
+}
+
+/**
+ * The starting annotated result object (before fetching)
+ */
+export const defaultBaseResult = {
+  fetching: false,
+  stale: false,
+  error: undefined,
+  data: undefined,
+  extensions: undefined,
+};
+
+/**
+ * Wonka fetching process:
+ *
+ * 1. EMIT: set fetching:true
+ * 2. fetch response(s)
+ * 3. EMIT: build a result from response (fetching:false)
+ */
+export function fetchProcess<Data, Variables>(
+  baseResult: AnnotatedOperationResult<Data, Variables>,
+  client: Client
+) {
+  /** fetching:true when starting */
+  const baseSource = pipe(fromValue({ ...baseResult, fetching: true }), share);
+
+  /** fetch via urql */
+  const urqlSource = pipe(
+    client.executeRequestOperation(baseResult.operation),
+
+    // build annotated result from base (set fetching:false)
+    map(response => ({ ...baseResult, ...response })),
+    share
+  );
+
+  // concat combines sources into a sequential source
+  const source = pipe(concat([baseSource, urqlSource]), share);
```

Comment:
I'm struggling with the unsubscribe.

when a query is run, it doesn't seem to emit a completion signal.  ex: if i do something like this in `queryStore.ts`:

```ts
pipe(
fetchProcess(...)
toPromise
)
```

the promise will never resolve.  i tried looking at the react bindings, but i'm not seeing a good way to teardown when urql is finished.
<img width="313" alt="Screen Shot 2022-05-18 at 9 26 32 PM" src="https://user-images.githubusercontent.com/2755958/169183415-c123b59e-f398-4fa3-81b6-188841010fd9.png">

Reply:
Sources from operations never complete automatically unless an exchange or other process in urql cancels them forcefully. Instead, once you subscribe, you'll indefinitely get updated results from the `Client`, which is how caching and cache updates work. Hence, `toPromise` is a tricky one. It'll wait until the _last_ result, however, there is no last result to wait for. So if you're using the raw sources you'll have to do `take(1)`.

I'll resolve this btw since it's unrelated to the current PR :v: